test/cql-pytest/test_cast_data.py

```diff
@@ -134,4 +139,16 @@ def test_cast_from_counter_to_varchar(cql, table2, cassandra_bug):
     cql.execute(f'UPDATE {table2} SET c = c + 1000 WHERE p = {p}')
     assert [("1000",)] == list(cql.execute(f"SELECT CAST(c AS varchar) FROM {table2} WHERE p={p}"))
 
+# Test casts from various types *to* counter type. This is a rather silly
+# operation - casting to "counter" doesn't make a real counter. It could
+# have been supported the same as casting to bigint, but Cassandra chose not
+# to support it and neither do we.
+# The only case that works is the do-nothing casting of a counter to counter,
+# and that case is already checked in test_cast_from_counter().
+def test_cast_to_counter(cql, table3):
+    p = unique_key_int()
+    for col in ['i', 'a', 'bi', 'b', 'bool', 'd', 'dec', 'db', 'dur', 'f', 'addr', 'si', 't', 'tim', 'ts', 'tu', 'ti', 'u', 'vc', 'vi']:
+        with pytest.raises(InvalidRequest, match='cannot be cast'):
+            cql.execute(f"SELECT CAST({col} AS counter) FROM {table3} WHERE p={p}")
+
```

Comment:
It's quite easy to implement these reverse casts (just add a few more lines to the cst-functions.cc, and more tests) but I had two reason not to do it:
1. Casting **to** a counter doesn't make sense... It doesn't make the result "a counter", it's just a 64-bit integer. You can't even assign the result to a counter-type column (because you can't assign a counter!). If you want a 64-bit integer, you can cast to bigint. I think there is no sensible use for casting to "counter".
2. Cassandra doesn't implement them, so nobody will need them for compatibility.

I tried to explain this in the commit message and/or comments, but perhaps not well enough.

Reply:
I understand that it's unlikely to be useful, but the operation makes sense, so ideally it should be possible to do it.
There might be a strange use case where it's needed. Being able to cast a counter into float also didn't look like an important feature at first, but it turned out that people need it.

Reply:
I think it's beyond "unlikely to be useful", I think it genuinely doesn't make sense - what will you do with a "counter" value after you cast something to it? You can't write the result back to a counter column. 

But more importantly (in my view) is the issue that Cassandra doesn't support it. The other direction, which I implemented, "people need it" for two reasons - first, it makes sense: You have a real counter column, and want to cast it to some other type (like string or double). Second, it works in Cassandra. So people who wrote their applications against Cassandra, and used whatever worked, were upset that their application no longer works against Scylla. 

I think that as long as Cassandra beats Scylla in https://db-engines.com/en/ranking (today Cassandra is down to 12, Scylla up to 67 - which is impressive but still not as popular as Cassandra), I think compatibility is still an important issue.

Reply:
It makes sense to me - counter is a type of integer, just like `long` or `tinyint`.
But yeah, it's probably not that useful.